tests/run-tests.py: Auto detect whether the target has threading and the GIL or not.

dpgeorge opened the pull request:
### Summary

This PR makes the following changes to the test runner and CI:
- detect if the target has the `_thread` module and, if so, add the threading tests to the set of tests to run
- if the target has threading, detect if the target has the GIL enabled or not, and use that information to exclude thread mutating tests if the GIL is disabled
- work around threading bug in `tests/extmod/select_poll_eintr.py` test
- make `tests/thread/stress_aes.py` test do less work on stackless threading builds, so it finishes before the test timeout limit
- increase test timeout for unix qemu test runs
- skip some threading tests on macOS and on unix qemu

Currently the information whether a port has the GIL is hard-coded: unix/PC targets and the rp2 port are considered to be GIL-less, and all other targets have the GIL enabled (if they have threading enabled).  With the change here, some code is run to detect the GIL.  That uses the fact that native code won't bounce the GIL and can hog it, and times how long code runs.

Overall the changes here mean that the threading tests are run in many more configurations:
- on the unix port with the native emitter
- on the unix port with single precision float
- on the unix port with stackless mode enabled
- on the unix port using clang as a compiler
- on unix qemu, the architectures MIPS, ARM and RISCV-64
- on macOS
 
### Testing

Tested locally on the unix port, esp32 and rp2.  Let's also see how the CI works.

### Trade-offs and Alternatives

This uses a tricky technique to determine GIL/non-GIL.  An alternative would be to add some indication in a function whether the GIL is enabled, eg `sys.implementation._thread_info`.  But that will increase code size.

dpgeorge committed:
tests/run-tests.py: Detect threading and automatically run thread tests.

When detecting the target platform, also check if it has threading and
whether the GIL is enabled or not (using the new attribute
`sys.implementation._thread`).  If threading is available, add the thread
tests to the set of tests to run (unless the set of tests is explicitly
given).

With this change, the unix port no longer needs to explicitly run the set
of thread tests, so that line has been removed from the Makefile.

This change will make sure thread tests are run with other testing
combinations.  In particular, thread tests are now run:
- on the unix port with the native emitter
- on macOS builds
- on unix qemu, the architectures MIPS, ARM and RISCV-64

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/select_poll_eintr.py: Pre-allocate global variables.

This is a workaround for the case where threading is enabled without a GIL.
In such a configuration, creating a new global variable is not atomic and
threads have race conditions resizing/accessing the global dict.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/thread/stress_aes.py: Reduce test time on PC targets.

This thread stress test is quite intensive and can run for a long time on
certain targets.  For example, builds with stackless enabled are slower
than non-stackless, and this test in stackless mode takes around 2 minutes
on the unix port of MicroPython with the existing parameters of
`n_thread=20` and `n_loop=5`.

It's not really necessary to test 20 threads at once, that's not really
going to test anything more than 10 at once.  So reduce the parameters to
keep the running time reasonable.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Increase timeout for unix qemu test runs.

The qemu emulation introduces enough overhead that the
`tests/thread/stress_aes.py` test overruns the default timeout.  So
increase it to allow this test to pass.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks @dlech for the review here.  After I initially submitted this PR and all the CI failed on things I couldn't test locally, the scope of the PR increased quite a bit... so there was a lot of back and forth getting it to work.

dpgeorge commented on tests/thread/stress_aes.py:
```diff
@@ -263,11 +263,30 @@ def thread_entry(n_loop):
+try:
+    import micropython
+
+    try:
+
+        def stackless():
+            pass
+
+        micropython.heap_lock()
+        stackless()
+        micropython.heap_unlock()
+    except RuntimeError:
+        micropython.heap_unlock()
+        is_stackless = True
```
You are right, it's probably better to be more focused like that.  I will change it.

(I was just copying existing code for this from eg `tests/micropython/heapalloc.py`.)

dpgeorge commented on tests/feature_check/thread_gil.py:
```diff
@@ -0,0 +1,41 @@
+busy()
+while not thread_done:
+    time.sleep(0)
+dt = time.ticks_diff(time.ticks_ms(), t0)
```
Yes, it could be flaky.  But from testing on bare-metal and PC, it looks reasonable.  The timing is usually either very close to T or 2*T (depending if GIL is disable or enabled).

If it fails due to timing errors (running for too long), it will incorrectly think the GIL is enabled when it is not.  Then it will run the mutate tests and they'll probably all fail.  In such a case we can just rerun CI...

If it gets to be a real issue, could increase the timing here, eg double it (I just don't want the auto detection to take too long, because it's run at the start of each run of `run-tests.py`.)

projectgus replied:
I have some similar concerns, and generally prefer the approach of explicitly adding this information in `sys.implementation` or similar. Suggest we keep this approach now, but keep the explicit option in mind if it turns out to be flaky.

projectgus replied:
I think it'd be useful to add. Especially if we don't add any `_thread` property on builds without threading support, then we avoid any code size hit on the smallest and simplest ports.

projectgus replied:
Much neater, I agree worth the extra size & complexity!

dpgeorge commented on ports/unix/mpthreadport.c:
```diff
@@ -56,7 +56,7 @@
 
 // This value seems to be about right for both 32-bit and 64-bit builds.
-#define THREAD_STACK_OVERFLOW_MARGIN (8192)
+#define THREAD_STACK_OVERFLOW_MARGIN (4 * 8192)
```
I'm currently trying to work out if this change is even needed.

dpgeorge replied:
This change is not needed, it doesn't fix the issue I thought it would fix.  So I took it out.

dpgeorge committed:
tools/ci.sh: Skip thread/stress_heap.py test on macOS test run.

This test passes sometimes and fails other times.  Eventually that should
be fixed, but for now just skip this test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Skip thread/stress_recurse.py on unix qemu test runs.

This test passes sometimes and fails other times.  Eventually that should
be fixed, but for now just skip this test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
py/modsys: Add sys.implementation._thread attribute.

This is useful to distinguish between GIL and non-GIL builds.

Signed-off-by: Damien George <damien@micropython.org>

projectgus commented on docs/library/sys.rst:
```diff
@@ -90,6 +92,11 @@ Constants
+   The *_thread* entry was added in version 1.26.0 and if it exists then the
+   target has the ``_thread`` module.  The target enables the GIL (global
+   interpreter lock) if this attribute equals ``"GIL"``, otherwise the attribute
+   is ``"no-GIL"`` and the target does not enable the GIL.
```
Do you think MicroPython will ever have a variant that is non-GIL but thread safe (ala CPython `--disable-gil`?)

If yes then maybe we could make the attribute values `GIL` and `unsafe` (or similar), keeping the door open for a `safe` value in some future release.

projectgus replied:
(Would mean tests which test for no-GIL will need to test `_thread != "GIL", I guess

dpgeorge replied:
> Do you think MicroPython will ever have a variant that is non-GIL but thread safe

Yes, that's a possibility.  But, I would suspect that we gradually evolve the current no-GIL mode to a more safer behaviour, rather than have separate options for GIL/unsafe-no-GIL/safe-no-GIL.

That said, I'm happy to change "no-GIL" to "unsafe".  Although that could scare users, but OTOH it might push them to read the docs to understand what "unsafe" actually means.

Is there another word that could better describe the situation, eg "racey" or "raceful"?

dpgeorge replied:
OK, I've now renamed "no-GIL" to "unsafe", and updated the docs.

Also tested again on esp32, rp2 and unix.

dpgeorge committed:
tools/ci.sh: Increase timeout for stackless clang test runs.

Stackless mode makes `tests/thread/stress_aes.py` slow, around 75 seconds
for this CI job (probably due to contention among the many threads for the
GC lock, to allocate frames for function calls).  So increase the timeout
to allow this test to pass.

Signed-off-by: Damien George <damien@micropython.org>
